### PR TITLE
Rerun kernel initialization after restarting kernel session

### PR DESCRIPTION
**Pull request by joyceerhl**
This PR fixes https://github.com/microsoft/vscode-jupyter/issues/7016 and https://github.com/microsoft/vscode-jupyter/issues/7086

### CODE REVIEW

File: src/client/datascience/jupyter/kernels/kernelExecution.ts [hunk 1]
```diff
@@ -268,11 +266,14 @@ export class KernelExecution implements IDisposable {
...
-    private async restartExecution(_document: NotebookDocument, session: IJupyterSession): Promise<void> {
+    private async restartExecution(notebook: INotebook): Promise<void> {
         // Just use the internal session. Pending cells should have been canceled by the caller
-        return session.restart(this.interruptTimeout).catch((exc) => {
+        await notebook.session.restart(this.interruptTimeout).catch((exc) => {
             traceWarning(`Error during restart: ${exc}`);
         });
+
+        // Reinitialize the kernel after a session restart
+        await notebook.runInitialSetup();
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelExecution.ts:276]: This was never required for Native notebooks, now we have code thats specific to interactive widow.
I think we should put an `if` condition for this and ensure this only runs for Intearctive window, as we have initialization for Native Notebooks already in the IKernel clas.

**Review by DonJayamanne — CHANGES_REQUESTED** [review]
Interactive window specific code should be gated (temporarily).
Once we move interactive widow to use IKernel then this code will not be necessary or we should move the initialization to the IKernel class.

File: src/client/datascience/jupyter/kernels/kernelExecution.ts (hunk 1 above)
Comment (joyceerhl) [reply src/client/datascience/jupyter/kernels/kernelExecution.ts:276]: This is required for native notebooks, and this is how restart used to work until 2 weeks ago. See https://github.com/microsoft/vscode-jupyter/pull/6826/files#diff-7d0d36cd88827937fc9458d80a807484935c9d2b5d63501a6ec98eb1e5351c16L178

Comment (DonJayamanne) [reply src/client/datascience/jupyter/kernels/kernelExecution.ts:276]: I don't see any old code that called `notebook.runInitialSetup`
That's new, Or an I missing something

Comment (joyceerhl) [reply src/client/datascience/jupyter/kernels/kernelExecution.ts:276]: That's a new method I added to expose the initialization logic which used to run as part of INotebook.restartKernel

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/jupyter/kernels/kernelExecution.ts (hunk 1 above)
Comment (DonJayamanne) [reply src/client/datascience/jupyter/kernels/kernelExecution.ts:276]: > PR is to move the initialization logic to IKernel.initializeAfterStart. What do you think?

Yes, thats much better, that way all the initialization is in one place.

**Review by DonJayamanne — APPROVED** [review]

**Review by IanMatthewHuff — APPROVED** [review]

**Comment by joyceerhl** [comment]
Closing as I think we want to rework how restart works and start a new kernel instance instead so that the INotebook / IKernel object doesn't need to remember whether it's been restarted.

**Pull request closed without merging**